jython-cli with examples

wfouche opened the pull request:
Cloned from python-cli@wfouche

wfouche commented:
@Stewori , @jeff5 - this pull request is ready to be reviewed.

wfouche reviewed (commented):
Ready to be reviewed.

Stewori commented:
FWIW, to review `python_cli.java` I just added the libs to apidia:
https://apidia.net/mvn/org.tomlj/tomlj
https://apidia.net/mvn/dev.jbang/jash

Stewori commented on jython_cli.java:
```diff
@@ -0,0 +1,210 @@
+
+public class jython_cli {
+
+    // Increment this number with each new release of python-cli
```
You mean "release of jython-cli", right?

wfouche replied:
Correct. Fixed

Stewori commented on jython_cli.java:
```diff
@@ -0,0 +1,210 @@
+//DEPS dev.jbang:jash:0.0.3
+//DEPS org.tomlj:tomlj:1.1.1
+//DEPS org.python:jython-standalone:2.7.4
+//JAVA 21
```
The semantics of these version numbers is maven semantics, right? And maven has a special syntax for pinning, which is not used here, so effectively this means, >=2.7.4, >=21, etc. I suppose. That would be convenient, since we would not have to remember updating this for every Jython release.

wfouche replied:
The >=2.7.4 version format is straight from the Python PEP-723 inline script metadata format in which version numbers are specified like this.  I wanted to align with PEP-723.  PEP-723 also support == which is version pinning, so I could have written this as 

* ==2.7.4
* ==21

which is  more appropriate (for now) and mirrors what is really implemented. Currently, I just take the version number and use  it as is.

On second thought, it is better to align with Maven and just use the numbers without operators because the operator functionality is not supported.

* 2.7.4
* 21

Will update the code accordingly.

Stewori replied:
The provided values look like maven coordinates, which is why I assumed the version numbers would be passed through to maven at some point. Taking a fresh look at https://maven.apache.org/pom.html#Dependency_Version_Requirement_Specification tells me that the "soft" interpretation of version numbers without further decoration is not as soft as I thought. In particular, I hoped it would be possible to write just "2.7", but that appears to work only in a version range. "[2.7,)" would be what I find ideal. It would free us from the burden of updating here. Normally, this is not recommended for sake of build reproducability. However, this use case is not for building and I want to think that Jython's 2.7.x train won't introduce bigger breakages but rather tiny fixes. OTOH one cannot really know and we might want to avoid introducing maven-specific version syntax. So, perhaps it's still best to stick with 2.7.4.

jeff5 replied:
I don't think the version numbers have anything to do with PEP-723, which only defines framing in a Python source. Isn't this being interpreted by JBang? It doesn't tell us the semantics or where it is borrowing from.

wfouche replied:
The version numbers are used to create `//DEPS` definitions in the JBang Java file, and have reffer to Maven Central artifacts.

Stewori commented on jython_cli.java:
```diff
@@ -0,0 +1,210 @@
+    public static void main(String[] args) throws IOException {
+        List<String> deps = new ArrayList<>();
+        String jythonVersion = "2.7.4";
+        String javaVersion = "21";
```
Is it really necessary to have the version numbers redundantly in multiple places as hard-coded constants? I acknowledge that it would be overly complicated to read them from the comments and this might be the only way feasible with reasonable effort. Just asking to confirm there is no better way. Repeating the same constant at multiple places is asking for errors on update.

wfouche replied:
Unfortunately, we need to duplicate this information as there is no easy way to read the //DEPS information from the Java file.

Stewori commented on jbang-catalog.json:
```diff
@@ -0,0 +1,8 @@
+  "aliases": {
+    "jython-cli": {
+      "script-ref": "jython_cli.java",
+      "description": "Run Jython 2.7.4"
```
Just a thought: Maybe we can just write 2.7 so this description would not require frequent updating.

wfouche replied:
Fixed.

Stewori commented:
@jeff5 what is your opinion on this repo license-wise? Similar repos of other projects often come without a license at all. We could put it under Jython's usual license, but that feels like an overkill. Introducing MIT or Apache license to the org OTOH does also feel unnecessary. Can we just leave it as it is without license? What would it take to put it under Jython's license? Would it be enough to copy the LICENSE file from Jython's main repo?

wfouche commented:
@Stewori , I've made the necessary code changes in reponse to your review comments.

Stewori commented on README.md:
````diff
@@ -1,2 +1,180 @@
+[Jython metadata for JBang]
+```
+# /// jbang
+# requires-jython = ">=2.7.4"
````
Hmm this contradicts the earlier description "2.7.1 ... 2.7.4". Either make this ">=2.7.1" or change the other one to  "2.7.4 ..." (open-ended as I suggested above) , depending on whether you think 2.7.1-2.7.3 should be considered viable options.

wfouche replied:
Fixed

Stewori commented on README.md:
````diff
@@ -1,2 +1,180 @@
+```
+
+```
+2.7.4.0
````
Okay, I see. This script is Jython 2.7.4-specific and has a nano-version component to track the script version. Does this mean when Jython 2.7.5 is released and let's say this script got multiple iterations, whould after 2.7.4.23 come 2.7.5.0 or 2.7.5.24? Perhaps this version semantics should be explained for future maintainers.

wfouche replied:
> 2.7.4.23 come 2.7.5.0

When the Jython version number changes, the FIX_NUMBER is reset to 0. Will update the documentation accordingly.

wfouche replied:
Fixed.

wfouche commented on README.md:
```diff
@@ -1,2 +1,180 @@
+
+`jython-cli` uses the `PEP-723` inline script metadata format to specify:
+
+* The version of Jython to use: 2.7.1, ... 2.7.4
```
Fixed

Stewori commented on jython_cli.java:
```diff
@@ -19,6 +19,8 @@
 
     // Increment this number with each new release of jython-cli
+    // But if the version number of Jython changes (for example 2.7.4 becomes 2.7.5), then
+    // the FIX_NUMBER is reset to 0 again.
```
Please add one more sentence describing the role of this number for the catalog version, like:
```
// This number appended to the Jython version forms the version of jython-cli as
// [Jython version].[FIX_NUMBER], e.g. 2.7.4.0
```

wfouche replied:
Fixed. Thanks.

Stewori commented on README.md:
```diff
@@ -1,2 +1,181 @@
+
+`jython-cli` uses the `PEP-723` inline script metadata format to specify:
+
+* The version of Jython to use: ..., 2.7.2, 2.7.3, 2.7.4, ...
```
This is an interesting take. I rather expected either

`* The version of Jython to use: 2.7.1, ...`

or

`* The version of Jython to use: 2.7.4, ...`

(depending on how much you would recommend against pre-4 2.7 versions).
Why so complicated?

wfouche replied:
> The version of Jython to use: 2.7.1, ...

Fixed.

Stewori commented:
Okay, I think my recent comments are the final ones. Will give @jeff5 some time to review if he wants.

jeff5 commented on jython_cli.java:
```diff
@@ -0,0 +1,212 @@
+        String javaVersion = "21";
+        String javaRuntimeOptions = "";
+        String jbangIntegrations = "true";
+        String ls = System.lineSeparator();
```
I don't usually have any problems on Windows just using `\n` for lines. It's not like we're reading the text with notepad.

wfouche replied:
Fixed.

jeff5 commented on jython_cli.java:
```diff
@@ -0,0 +1,212 @@
+        String dep = "org.python:jython-standalone:" + jythonVersion;
+        deps.add(dep);
+
+        byte[] data = Files.readAllBytes(Paths.get(scriptFilename));
```
Well this is a surprise ... so here we write a new Java program that is specific to the original Python script, and contains some stuff from the JBang directives in the Python. When run, with JBang, it invokes `PythonInterpreter`.

This seems very indirect. I think I was expecting argument to this tool either to go directly to `o.p.util.jython.main()` or for Java to execute the JAR with those.

I realise this surprise comes because nothing so far has explained what this program actually does, and how it goes about it.

wfouche replied:
@jeff5 , I thought about your comment and realized the added complexity was really a left-over from code I wrote to embed a Jython script as a reporting tool in a larger Java/Kotlin application and that this functionality is not needed this scenario. I removed the script caching code and clean-up the Java code. It now ends up  just calling jython.main(args). I'll explain the program logic in a follow up comment.

jeff5 reviewed (changes requested):
Something to explain what this does and why it works this way is essential, I think. Javadoc might be a good way.

Stewori commented on jbang-catalog.json:
```diff
@@ -0,0 +1,8 @@
+  "aliases": {
+    "jython-cli": {
+      "script-ref": "jython_cli.java",
+      "description": "Run Jython 2.7"
```
Are you suggesting to call this `jython2-cli`? Or a future one `jython3-cli`? I think there is no point in thinking about Jython 2.5.x because it's obsolete more than 10 years by now. So we kind of have the luxurious situation that we only have to consider micro version updates and can assume the 2.7 to be rather stable. My impression is that this issue is a rather academic one since Jython has not nearly as many users as CPython (sometimes I wonder whether there are any left), so the impact would be much smaller. We can be glad if some adoption of this script will take place at all. Then, it's somewhat speculative when and if a Jython 3 will be released to trouble us here. Do you think this is reason enough to call this script `jython2-cli`? I kind of suspect it will be the least of our troubles. All in all, I would not be concerned too much and probably replicate CPython's way *iff* an actual issue would really manifest (that this script does actually have non-zero users that is). Also, ideal or not, people are used to CPython's handling of the issue, so it is easily explained if we copy their approach.
My suggestion would be:
- take this script as it is more or less
- on release of Jython 3.0.0-alpha1, add `jython3-cli`, conserve the current one in `jython2-cli` and adjust `jython-cli` to issue a warning that in a couple of years it will default to running Jython 3.

Or what about this one: Try with Jython 3, catch the parser exception and fallback to Jython 2 (and issue a warning).

jeff5 replied:
Quite like advancing `jython-cli` to latest always and introducing either a more specifically named script as needed, or a version argument.

Stewori replied:
"latest always" across major version would cause a sudden breakage in case of Jython/Python 2-3. Encoding some version arg into the script name is an interesting idea, but we don't have a spec for it and this concept is unprecedented in the Python world. Version argument would cause a sudden breakage when the default value is changed from a 2.x to a 3.x version.

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,212 @@
+
+//DEPS dev.jbang:jash:0.0.3
+//DEPS org.tomlj:tomlj:1.1.1
+//DEPS org.python:jython-standalone:2.7.4
```
Fixed. Now using jython-slim:2.7.4

wfouche commented:
> I think if @wfouche lets GitHub add a license file, and it is for one of the PSF-specified ones, we are halfway there.

File LICENCE.txt added. It was extracted from the Jython LICENSE.txt file, and only represents a portion of that file.

wfouche commented:
I'm still need to review all the new comments and questions and respond to them, and will post a comment here when I have done so.

Stewori commented:
Is it really necessary to introduce yet another doc format (AsciiDoc, IMPLEMENTATION.adoc) into our setup? So far it's only ~200 lines. Perhaps markdown would be more compatible to what we are using elsewhere (homepage, readme). We already have to deal with rst vs markdown occasionally. If this is asked too much I would volunteer to convert those existing ~200 lines to markdown (under the condition that future doc is added in markdown readily).

wfouche commented:
> Is it really necessary to introduce yet another doc format

I've been using Asciidoc for a very long time so it just very natural for me to write documentation in IntelliJ or VS Code using Asciidoc.

No problem. I'll convert the document to an MD file.

wfouche commented:
I used `downdoc` to convert the .adoc file to a .md file:

* https://github.com/opendevise/downdoc?tab=readme-ov-file#how-does-it-work

Stewori commented:
Thanks for converting to markdown!

wfouche commented:
> I'm still need to review all the new comments and questions and respond to them, and will post a comment here when I have done so.

It took a while, but I think I'm done with all the requested updates. File IMPLEMENTATION.md describes what jython-cli does. File jython-cli.java were updated to simplify the implementation  and also make it more robust.

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,212 @@
+        String javaFilename = javaClassname + ".java";
+
+        // Parse PEP 723 text block
+        {
```
"Poor man's lambgda function" ...

I fixed it.

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,212 @@
+                    if (line.startsWith("# /// jbang")) {
+                        found = true;
+                    }
+                    else if (line.startsWith("# ///")) {
```
I beefed up the parsing the the metadata block.

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,212 @@
+
+        // Parse PEP 723 text block
+        {
+            StringBuffer tomlText = new StringBuffer("");
```
Fixed

jeff5 commented:
> @jeff5 , @Stewori - please do a final review. I've responded to all the change requests.

Will do. I've been saving observations for your signal. But thanks for resiling from asciidoc, which would have been one of them.

Stewori commented on LICENSE.txt:
```diff
@@ -0,0 +1,47 @@
+
+1. This LICENSE AGREEMENT is between the Python Software Foundation
+("PSF"), and the Individual or Organization ("Licensee") accessing and
+otherwise using this software ("Jython") in source or binary form and
```
Since this license refers to the repo at hand, `Jython` should be replaced by `jython-cli` I guess. To make it clear it is about the cli specifically. Jython itself is of no concern here as it has its own license copy in its own repo.

Stewori replied:
(search/replace alls occurrences in this document of course)

wfouche replied:
Renamed Jython to jython-cli

Stewori commented on jython_cli.java:
```diff
@@ -0,0 +1,226 @@
+        String ext = System.getProperty("os.name").toLowerCase().startsWith("win") ? ".cmd" : "";
+        String[] jargs = params.toString().split("\\s+");
+        try (Stream<String> ps = Jash.start("jbang" + ext, jargs).stream()) {
+                ps.forEach(System.out::println);
```
double indentation?

wfouche replied:
Fixed

Stewori commented on jython_cli.java:
```diff
@@ -0,0 +1,226 @@
+    public static void main(String[] args) throws IOException {
+        List<String> deps = new ArrayList<>();
+        String jythonVersion = DEFAULT_JYTHON_VERSION;
+        String javaVersion = DEFAULT_JAVA_VERSION;
```
What about Jeff's idea to get the versions from Java and Jython?
`System.getProperty("java.version")` / perhaps better [`Runtime.feature()`](https://apidia.net/java/OpenJDK/24/?pck=java.lang&cls=.Runtime#Version.feature-)
[`org.python.Version.PY_VERSION`](https://apidia.net/java/Jython/2.7.4/?pck=org.python&cls=.Version#PY_VERSION)

wfouche commented:
@Stewori , I'm done with your requested code changes. Thanks for the feedback. 

@jeff5 , over to you.

wfouche commented:
@jeff5 , fixing the Java and Jython default version related issues that you pointed out (and @Stewori reminded me of), I think I have addressed all the issues you have raised.   Please confirm when you have time.

jeff5 commented on jython_cli.java:
```diff
@@ -0,0 +1,245 @@
+            "public class __CLASSNAME__ {",
+            "",
+            "    public static void main(String[] args) {",
+            "        System.setProperty(\"python.console.encoding\", \"UTF-8\");",
```
This forcing of Jython to believe it is in a UTF-8 console is not entirely correct, but it is close enough that I am in two minds about it. I spent quite some time exploring this. If `python.console.encoding` is not defined we try to work it out from properties the JVM might supply. Oracle JVMs do this up to at least 17, but the Temurin ones do not. At 21, the information is in a standard place in both implementations, although we don't currently look there :( because I only just found that out. See https://github.com/jython/jython/issues/384.

If you output characters to a console that can't print them Python (including Jython) raises an encoding error. But if you tell them the `sys.stdout.encoding` is UTF-8, as you effectively have here, Python knows no better than to pass them on to Java which quietly substitutes `?`.

Our output of text is actually a nightmarish logical inversion of what you might expect because Python 2 output is fundamentally done in bytes while Java expects characters. So the claimed _encoding_ is actually used to _decode_ the bytes `sys.stdout` receives to characters, so that Java can put the characters on the console using the encoding it knows really applies there.

The bytes you give `sys.stdout` may well be the result of _encoding_ a `unicode` string using the claimed encoding and it doesn't matter that this is wrong because we'll use the same encoding to turn them back into characters. Or rather it only matters if you think you should get the error CPython would give you.

Bottom line: I think we should leave this as it is and document it as a known off-spec behaviour, until we find a stronger reason to correct it.

jeff5 reviewed (changes requested):
This is great! It's going to be very useful. I have very few functional issues to raise (and only one looks difficult).

My non-functional issues are towards maintainability: clarity of specification of the TOML and easy comprehension by someone who wants to make a change later.

There will be the PSF contributor agreement to sign, too.

jeff5 commented on IMPLEMENTATION.md:
```diff
@@ -0,0 +1,216 @@
+
+The purpose of the `jython-cli` program is to generate a Jython.java type program from dependency information (metadata) added to a Jython script.
+
+## PEP-723 (Inline Script Metadata)
```
I tested the "scans & ignores" properties and we seem to be good.

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,245 @@
+        }
+
+        // --version
+        if (args.length == 1 && args[0].equals("--version")) {
```
Fixed.

Renamed `--version` to `--jython-cli-version`.

Output from the various version options:

```bash
$ jbang run jython-cli -V
Jython 2.7.4
$ jbang run jython-cli --version
Jython 2.7.4
$ jbang run jython-cli --jython-cli-version
jython-cli 2.7.4.0
```

wfouche commented on IMPLEMENTATION.md:
```diff
@@ -0,0 +1,216 @@
+# jython-cli
+
+Article [Running Jython scripts with JBang using a Java helper program](https://medium.com/@werner.fouche/running-jython-scripts-with-jbang-using-a-java-helper-program-9ab9f8e35ddc) describe how a **Jython.java** helper program is used to resolve the Maven dependencies of Jython script **banner.py**.
```
Fixed

wfouche commented on IMPLEMENTATION.md:
````diff
@@ -0,0 +1,216 @@
+        |___/
+```
+
+The purpose of the `jython-cli` program is to generate a Jython.java type program from dependency information (metadata) added to a Jython script.
````
Fixed.

wfouche commented:
@jeff5, thanks for the very detailed and constructive feedback. It will take me a days to address all the issues. I'll post a message here when I've resolved them all.

wfouche commented:
@jeff5, It took me a few days to consider how to resolve the remaining issues, and eventually I decided to use a diffenent but consistent approach to running Jython scripts, one that allows console input to be available when it should be. The revised implementation is uncomplicated and it only provides the minimum required amount of functionality. All Jython command-line options remain accessible and are not shadowed. 

```bash
 jbang run jython-cli -i examples/banner.py 
['examples/banner.py']
      _       _   _                   ____   _____ 
     | |_   _| |_| |__   ___  _ __   |___ \ |___  |
  _  | | | | | __| '_ \ / _ \| '_ \    __) |   / / 
 | |_| | |_| | |_| | | | (_) | | | |  / __/ _ / /  
  \___/ \__, |\__|_| |_|\___/|_| |_| |_____(_)_/   
        |___/                                      
>>> print 1+2
3
>>> 
```

The new implementation is contained in class JythonCli.java. The previous (research quality code) class jython_cli.java is gone. The IMPLEMENTATION.md guide describes the new implementation details. The metadata has also been simplified.

The code is ready to be reviewed.

wfouche commented:
> This is getting much clearer, thanks, including some things I thought might be good but didn't ask for. I'll hold off until you say go.

I'm done polishing. I made some last minute changes yesterday while I was running additional tests. 

So, the code is ready to be reviewed.

wfouche commented:
I see I'm still using the standalone JAR when calling JBang at the end of the script. 

![image](https://github.com/user-attachments/assets/958f30aa-70b3-4bdc-ba43-f8bcddde4463)

I'm submitting a fix for this issue now.

wfouche commented:
Done.

jeff5 commented:
@wfouche This is looking good. We need you to sign this and confirm back here that you have. https://www.python.org/psf/contrib/contrib-form/

wfouche commented on JythonCli.java:
Indenting changed from 2 column to 4 columns.

javadoc comments added to the main methods and global shared variables.

wfouche commented on JythonCli.java:
```diff
@@ -0,0 +1,158 @@
+/// usr/bin/env jbang "$0" "$@" ; exit $?
+
+//DEPS org.tomlj:tomlj:1.1.1
+//DEPS org.python:jython-slim:2.7.4
```
Good catch! It is not needed. Removed.

wfouche commented:
@jeff5 , I've made the requested changes and signed the contributor agreement, so it is all done. Over to you. (smile)

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,245 @@
+///usr/bin/env jbang "$0" "$@" ; exit $?
```
Re-implemented as JavaCli.java, now better structured and easy to follow program logic.

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,245 @@
+    // FIX_NUMBER is reset to 0 again, e.g. 2.7.5.0
+    private static final int FIX_NUMBER = 0;
+
+    public static String getJvmMajorVersion() {
```
Re-implemented as JythonCli.java

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,245 @@
+        }
+        String ext = System.getProperty("os.name").toLowerCase().startsWith("win") ? ".cmd" : "";
+        String[] jargs = params.toString().split("\\s+");
+        try (Stream<String> ps = Jash.start("jbang" + ext, jargs).stream()) {
```
Resolved in new JythonCli.java implementation.

wfouche commented on jython_cli.java:
```diff
@@ -0,0 +1,245 @@
+//DEPS org.tomlj:tomlj:1.1.1
+//DEPS org.python:jython-slim:2.7.4
+
+// jython-cli can be run with Java 8 from the command-line as follows:
```
Fixed.

wfouche commented:
@Stewori  , if we do not hear from @jeff5 this week , are you able to merge this pull request since @jeff5 was happy to accept the code in its current state (given that the formatting is changed to 4 spaces for indents and that I sign the Python Contributor agreement); both issues have been taken resolved.

jeff5 commented on IMPLEMENTATION.md:
That definitely solves the overlap with README.md _I think_ the "Metadata" heading in the README covers the need for a user guide, as there is not much one can write in the "/// jbang" section.

jeff5 reviewed (approved):
I'm happy to see this merged. I suggest a squash merge so we do not keep all the twists and turns of the journey.

I've raised issues for the two things bothering me, as they don't stop it working. The formatting is a bit wonky, but that's one click to fix in the IDE.

Thanks for your patient work @wfouche !

jeff5 commented:
@wfouche: When we merge to ` main` your repo on GitHub will be out of sync as it ` main ` will be nothing like the one here, and any further work on it will just deepen the confusion.

I think the answer may be to rename your ` main` and sync again to a new `main` that is your default branch. (See if you get a `main` the same as ours.) Now I describe it, your better option is probably to make a fresh clone.

The way to avoid this, when you come to contribute again, is to make a branch to represent your planned change, then at some point raise a PR from that. We follow this pattern (but don't have the same automation): https://devguide.python.org/getting-started/pull-request-lifecycle/#quick-guide
